### PR TITLE
ADR 005: Observability

### PR DESCRIPTION
**Pull request by bastbu**
# PR for issue #468 

## What is being addressed

ADR 004 on observability

### CODE REVIEW

File: Docs/ArchitectureDecisions/004_observability [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+## Status
+
+Accepted
+
+## Context
+
+The goal of observability for the LoRaWAN IoT Edge starter kit is to:
+
+1. Monitor if the LoRaWAN Starter Kit solution works accoridng to the user expectations regarding the following factors:
```
Comment (roel4ez) [inline Docs/ArchitectureDecisions/004_observability:15]:
```suggestion
1. Monitor if the LoRaWAN Starter Kit solution works according to the user expectations regarding the following factors:
```

File: Docs/ArchitectureDecisions/004_observability [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+   - Coverage. The data is coming from the majority of observed IoT assets
+   - Freshness. The data coming from the assets is fresh and relevant
+   - Throughput. The data is delivered from the assets without significant delays.
+   - Correctness. The ratio of erors and lost messages from the assets is small
```
Comment (roel4ez) [inline Docs/ArchitectureDecisions/004_observability:19]:
```suggestion
   - Correctness. The ratio of errors and lost messages from the assets is small
```

**Review by roel4ez — COMMENTED** [review]
- [ ] The file does not have an extension

Also, RE OpenTelemetry: .NET6 will have an implementation of the Metrics API: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md

File: Docs/ArchitectureDecisions/004_observability.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+## Decisions
+
+We will support Azure Monitor as a first-class monitoring solution for our starter kit. A user can opt-in to use Application Insights with the starter kit, in which case we will support a rich set of observability features. If the user decides to not use Application Insights, we will still support essential monitoring capabilities. This means that we will:
+
+- Track LNS logs in Application Insights (when opted in). We will adhere to the [IoT Edge recommended format](https://docs.microsoft.com/en-gb/azure/iot-edge/how-to-retrieve-iot-edge-logs?view=iotedge-2020-11#recommended-logging-format) for the structure of the log console output. Export of logs to anything else than Application Insights requires a custom solution by the user and is not supported by the starter kit.
+- Always expose metrics using [prometheus-net](https://github.com/prometheus-net/prometheus-net).
+  - Additionally, we track LNS metrics using the ASP.NET Core Application Insights SDK (when opted in)
+- Track traces using the Application Insights SDK (when opted in)
+- Support alerts when using Application Insights
```
Comment (eedorenko) [inline Docs/ArchitectureDecisions/004_observability.md:33]: Support alerts when using Application Insights and/or Log Analytics (with Prometheus format and metrics collector module)

File: Docs/ArchitectureDecisions/004_observability.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+### Traces
+
+We use built-in tracing from Azure Application Insights (on an opt-in basis). This works well for function calls and correlation to other services, such as Key Vault. We will not include message flow end to end tracing for now, but will reevaluate with [#695](https://github.com/Azure/iotedge-lorawan-starterkit/issues/695).
+
+### Metrics
+
+The core modules `edgeHub` and `edgeAgent` support emitting metrics through a Prometheus endpoint, using the strategy described in [Access built-in metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-built-in-metrics?view=iotedge-2020-11). To collect these metrics and integrate everything with Application Insights, we use the metric collector (preview) as suggested in [Collect and transport metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-collect-and-transport-metrics?view=iotedge-2020-11&tabs=iothub) to export metrics to a Log Analytics storage.
+
```
Comment (eedorenko) [inline Docs/ArchitectureDecisions/004_observability.md:59]: "integrate everything with Application Insights"? perhaps "integrate everything with Azure Monitor"

File: Docs/ArchitectureDecisions/004_observability.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+- Eliminates any dependencies on Azure Monitor services (Log Analytics / Application Insights) for essential monitoring
+- Potentially gives ability to work offline if metrics are sent by the collector module through the Edge Hub using device-to-cloud channel.
+- It's up to the customer to configure how, where and what metrics to deliver from any module on an edge device.
+
+In addition to this, we will support Application Insights metrics on an opt-in basis. When enabled, we will deliver most metrics (custom and default from LNS, except the edgeAgent and edgeHub metrics, which can only be delivered to Log Analytics) to Application Insights. This will ensure that we get many of the features that we get with Application Insights out of the box (Live Metrics, integration with alerts and workbooks), while still keeping the flexibility of consuming the metrics in Prometheus format and all the advantages that come with it. This comes at the cost of increased implementation complexity.
+
```
Comment (eedorenko) [inline Docs/ArchitectureDecisions/004_observability.md:70]: Perhaps it's worth mentioning here that IoT Hub comes with [curated workbooks](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-explore-curated-visualizations?view=iotedge-2020-11&tabs=devices%2Cmessaging) and a [bunch of predefined queries for alerts](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-create-alerts?view=iotedge-2020-11) based on built-in Prometheus-format metrics that are delivered by the metrics collector module to LogAnalytics.

File: Docs/ArchitectureDecisions/004_observability.md (hunk 3 above)
Comment (bastbu) [reply Docs/ArchitectureDecisions/004_observability.md]: Done with 0473b66eea2406aa0a9b1cc346e8891dc587f6c0

File: Docs/ArchitectureDecisions/004_observability.md (hunk 5 above)
Comment (bastbu) [reply Docs/ArchitectureDecisions/004_observability.md:70]: Addressed this with 495aabfa8cd555f40a1bf5e3607a9628a68c6f8b

File: Docs/ArchitectureDecisions/004_observability.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+### Metrics
+
+The core modules `edgeHub` and `edgeAgent` support emitting metrics through a Prometheus endpoint, using the strategy described in [Access built-in metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-built-in-metrics?view=iotedge-2020-11). To collect these metrics and integrate everything with Azure Monitor, we use the metric collector (preview) as suggested in [Collect and transport metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-collect-and-transport-metrics?view=iotedge-2020-11&tabs=iothub) to export metrics to a Log Analytics storage.
+
+We will always expose LNS custom metrics in Prometheus format using [prometheus-net/prometheus-net](https://github.com/prometheus-net/prometheus-net), such that they can be consumed by any scraper that supports the Prometheus format. This will give us the following features:
+
+- Unified metrics format accross all modules in the Edge device.
+  - The Prometheus format is industrial standard understood by various consumers.
```
Comment (danigian) [inline Docs/ArchitectureDecisions/004_observability.md:63]: nit: Is this a sub-item of the list?

File: Docs/ArchitectureDecisions/004_observability.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+| DeviceLoadRequests         | Number of device load requests                               | LNS    | LoRaWan   | Gateway Id                 |
+| JoinRequests               | Number of join requests                                      | LNS    | LoRaWan   | Gateway Id                 |
+| D2CMessagesReceived        | Number of messages received from device                      | LNS    | LoRaWan   | Gateway Id, Device Id      |
+| D2CMessagesDelivered       | Number of messages sent to upstream                          | LNS    | LoRaWan   | Gateway Id, Device Id, To  |
+| D2CMessagesError           | Number of errors in sending messages to upstream             | LNS    | LoRaWan   | Gateway Id, Device Id, To  |
+| D2CMessagesProcessingError | Number of errors processing (decoding, decrypting) messages  | LNS    | LoRaWan   | Gateway Id, Device Id, To  |
+| D2CMessageSize             | Message size in bytes received from device                   | LNS    | LoRaWan   | Gateway Id, Device Id      |
+| D2CMessageSizeUpstream     | Message size in bytes sent upstream                          | LNS    | LoRaWan   | Gateway Id, Device Id, To  |
+
```
Comment (danigian) [inline Docs/ArchitectureDecisions/004_observability.md:89]: What about C2D related metrics? Is there a plan for those?

**Review by danigian — APPROVED** [review]

File: Docs/ArchitectureDecisions/004_observability.md (hunk 6 above)
Comment (bastbu) [reply Docs/ArchitectureDecisions/004_observability.md:63]: Yes, it was intended as sub-item

File: Docs/ArchitectureDecisions/004_observability.md (hunk 7 above)
Comment (bastbu) [reply Docs/ArchitectureDecisions/004_observability.md]: Added the comment that also for C2D we should add these metrics

**Pull request merged**